pytensor/tensor/subtensor.py

```diff
@@ -178,16 +215,51 @@ def as_index_literal(
             as_index_literal(idx.step),
         )
 
+    if not isinstance(idx, Variable):
+        raise NotScalarConstantError()
+
+    if isinstance(idx.type, NoneTypeT):
+        return None
+
+    if isinstance(idx, ScalarConstant):
+        return cast(int | None, idx.data)
```

Comment:
A ScalarConstant can't be None
```suggestion
        return cast(int, idx.data)
```